File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        return solve(nums1, nums1.size(), nums2, nums2.size());
    }
private:
    double solve(vector<int>& nums1, int n, vector<int>& nums2, int m) {
        
        if (n > m) return solve(nums2, m, nums1, n);
        
        int low = 0;
        int high = n;
        int partSize = (n + m + 1) / 2;
        
        while (low <= high) {
            int partX = (low + high) / 2;
            int partY = partSize - partX;
            
            int maxLeftX = partX == 0 ? INT_MIN : nums1[partX - 1];
            int minRightX = partX == n ? INT_MAX : nums1[partX];
            
            int maxLeftY = partY == 0 ? INT_MIN : nums2[partY - 1];
            int minRightY = partY == m ? INT_MAX : nums2[partY];
            
            if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
                if ((m + n) % 2 == 0) {
                    return (double(max(maxLeftX, maxLeftY) + double(min(minRightX, minRightY)))/2);
                } else {
                    return double(max(maxLeftX, maxLeftY));
                }
            } else if (maxLeftX > minRightY) 
                high = partX - 1;
            else {
                low = partX + 1;
            }
            
        }
        
        return double(0);
    } 
};

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:

        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1
        
        total = len(nums1) + len(nums2)
        left_size = total // 2

        l = 0
        r = len(nums1) - 1

        while True:
            mid1 = (l + r) // 2
            right_idx = left_size - (mid1 + 1) - 1
            
            left_num1 = nums1[mid1] if mid1 >= 0 else float("-inf")
            right_num1 = nums1[mid1 + 1] if mid1 + 1 < len(nums1) else float("inf")
            left_num2 = nums2[right_idx] if right_idx >= 0 else float("-inf")
            right_num2 = nums2[right_idx + 1] if right_idx + 1 < len(nums2) else float("inf")

            if left_num1 <= right_num2 and left_num2 <= right_num1:
                if total % 2 == 0:
                    return (max(left_num1, left_num2) + min(right_num1, right_num2)) / 2
                else:
                    return min(right_num1, right_num2)

            if left_num1 > right_num2:
                r = mid1 - 1
            else:
                l = mid1 + 1





        